crates/flox-rust-sdk/src/models/environment/generations.rs

```diff
@@ -0,0 +1,360 @@
+use serde_with::{DeserializeFromStr, SerializeDisplay};
+use thiserror::Error;
+
+use super::path_environment::{Original, PathEnvironment};
+use super::{copy_dir_recursive, PathPointer, ENV_DIR_NAME};
+use crate::providers::git::{GitCommandProvider, GitProvider};
+
+const GENERATIONS_METADATA_FILE: &str = "metadata.json";
+
+pub struct ReadOnly {}
+
+pub struct ReadWrite {
+    /// A base directory that contains the tempdir
+    /// of the checked out generations branch
+    /// as well as tempdirs created by [PathEnvironment] transactions.
+    tempdir_base: PathBuf,
+}
+
+/// A representation of the generations of an environment
+///
+/// Example File layout:
+///
+/// ./
+/// ├── 1
+/// │  └── env
+/// │     ├── manifest.toml
+/// │     └── manifest.lock
+/// ├── 2
+/// │  └── env
+/// │    └── manifest.toml (lockfile is optional)
+/// ├── ... N
+/// │  └── env
+/// │     └── manifest.toml
+/// └── metadata.json
+pub struct Generations<State> {
+    /// A floxmeta repository/branch that contains the generations of an environment
+    ///
+    /// - When [ReadOnly], this is assumend to be bare, but it is not enforced.
+    /// - When [ReadWrite], this is required to not be bare.
+    ///   This is enforced when created using [Self::writable].
+    repo: GitCommandProvider,
+
+    /// A path pointer for the environment that will
+    /// be associated with a realized generation
+    pointer: PathPointer,
+
+    branch: String,
+
+    state: State,
+}
+
+impl<S> Generations<S> {
+    /// Read the generations metadata for an environment
+    pub fn metadata(&self) -> Result<Metadata, GenerationsError> {
+        read_metadata(&self.repo, &self.branch)
+    }
+}
+
+impl Generations<ReadOnly> {
+    /// Create a new generations instance
+    pub fn new(
+        environment_repr: GitCommandProvider,
+        ref_name: String,
+        pointer: PathPointer,
+    ) -> Self {
+        Self {
+            repo: environment_repr,
+            branch: ref_name,
+            pointer,
+            state: ReadOnly {},
+        }
+    }
+
+    pub fn writable(
+        self,
+        tempdir_base: PathBuf,
+    ) -> Result<Generations<ReadWrite>, GenerationsError> {
+        let repo = checkout_to_tempdir(
+            &self.repo,
+            &self.branch,
+            tempfile::tempdir_in(&tempdir_base).unwrap().into_path(),
+        )?;
+
+        Ok(Generations {
+            repo,
+            branch: self.branch,
+            pointer: self.pointer,
+            state: ReadWrite { tempdir_base },
+        })
+    }
+}
+
+impl Generations<ReadWrite> {
+    /// Realize a generation as a [PathEnvironment]
+    ///
+    /// The generation can then be safely modified
+    /// and registered as a new generation using [Self::add_generation].
+    pub fn realize_generation(
+        &self,
+        generation: usize,
+    ) -> Result<PathEnvironment<Original>, GenerationsError> {
+        let environment = PathEnvironment::new(
```

Comment:
question/suggestion blocking: since this is just returning a wrapper around an old generation, won't calling e.g. `install()` on the resulting `PathEnvironment<Original>` modify it in the git repo? I think that's not getting committed, but it seems a bit bug prone - can you comment or change that?

Reply:
I made a note in the comment
It's not getting committed, but it leaves the modified env behind.
You should not reuse the same `Generations` object to modify the same generation more than once.

This is a "regression" from eliminating one copy, to prevent this we would need to create a separate copy every time we (1) access an environment and (2) when we add an envrionment (this might be ptimized to a constant cost, but 🤷🏼‍♂️).

Personally I'd rather do a copy more and have it isolated, but we discussed _not_ doing that yesterday.
For now it's at least documented, we can come back to this if its itchy